contracts/ERC-721/ERC721Pool.sol

```diff
@@ -37,46 +67,66 @@ contract ERC721Pool is IERC721Pool, ERC20Wnft {
     /// PUBLIC NON-CONSTANT FUNCTIONS ///
 
     /// @inheritdoc IERC721Pool
-    function deposit(uint256[] calldata ids) external override {
-        if (ids.length == 0) {
-            revert ERC721Pool__InsufficientIn();
+    function deposit(uint256 id, address beneficiary) external override notFrozen {
+        // Checks: beneficiary is not zero address
+        if (beneficiary == address(0)) {
+            revert ERC721Pool__ZeroAddress();
         }
-        for (uint256 i; i < ids.length; ) {
-            uint256 id = ids[i];
-            require(holdings.add(id));
-            IERC721(asset).transferFrom(msg.sender, address(this), id);
-            unchecked {
-                ++i;
-            }
+        // Checks: Add a NFT to the holdings.
+        if (!holdings.add(id)) {
+            revert ERC721Pool__NFTAlreadyInPool(id);
         }
-        _mint(msg.sender, ids.length * 10**18);
-        emit Deposit(ids, msg.sender);
+
+        // Interactions: perform the Erc721 transfer from caller.
+        IERC721(asset).transferFrom(msg.sender, address(this), id);
+
+        // Effects: Mint an equivalent amount of pool tokens to the beneficiary.
+        _mint(beneficiary, 10**18);
+
+        emit Deposit(id, beneficiary, msg.sender);
     }
 
     /// @inheritdoc IERC721Pool
-    function withdraw(uint256[] calldata ids) public override {
-        if (ids.length == 0) {
-            revert ERC721Pool__InsufficientIn();
-        }
-        _burn(msg.sender, ids.length * 10**18);
-        for (uint256 i; i < ids.length; ) {
-            uint256 id = ids[i];
-            require(holdings.remove(id));
-            IERC721(asset).transferFrom(address(this), msg.sender, id);
-            unchecked {
-                ++i;
-            }
+    function rescueLastNFT(address to) external override onlyFactory {
+        // Checks: The pool must contain exactly one NFT.
+        if (holdings.length() != 1) {
+            revert ERC721Pool__MustContainExactlyOneNFT();
         }
-        emit Withdraw(ids, msg.sender);
+        uint256 lastNFT = holdings.at(0);
+
+        // Effects: Remove lastNFT from the holdings.
+        holdings.remove(lastNFT);
+
+        // Interactions: Transfer the NFT to the specified address.
+        IERC721(asset).transferFrom(address(this), to, lastNFT);
+
+        // Effects: Freeze the pool.
+        poolFrozen = true;
+
+        emit RescueLastNFT(lastNFT, to);
     }
 
     /// @inheritdoc IERC721Pool
-    function withdrawWithSignature(
-        uint256[] calldata ids,
-        uint256 deadline,
-        bytes memory signature
-    ) external override {
-        permitInternal(ids.length * 10**18, deadline, signature);
-        withdraw(ids);
+    function setENSName(address registrar, string memory name) external override onlyFactory returns (bytes32) {
+        bytes32 nodeHash = IReverseRegistrar(registrar).setName(name);
+        emit ENSNameSet(registrar, name, nodeHash);
```

Comment:
Are the `registrar` and `nodeHash` really necessary in this event ? Internally, the registrar call will fire a ` ReverseClaimed(address indexed addr, bytes32 indexed node)` event anyway [here](https://github.com/ensdomains/ens-contracts/blob/787c5d8f1a99ad14435a65784a7c5ceca1e2575e/contracts/reverseRegistrar/ReverseRegistrar.sol#L93).

Reply:
Thank @hroussille, I do agree that it is redundancy of emitting the registrar and nodeHash in the ENSNameSet event. 
I've made updates to the contracts In this [commit](https://github.com/hifi-finance/pooled-nft/pull/9/commits/8771a7df625605c1eab6f95836624a1b35ddb8e8) considering both @scorpion9979  and @hroussille [feedback](https://github.com/hifi-finance/pooled-nft/pull/9#discussion_r1197060854). To summarize:

1. I moved the ENSNameSet event from the ERC721Pool contract to the ERC721PoolFactory contract. This enables us to track the ENS names set for each pool created by the factory from a single source.
2. I've simplified the ENSNameSet event to emit only the pool and name parameters, as the registrar and nodeHash are already being emitted in the ReverseClaimed event from the IReverseRegistrar contract.
3. I have updated the mock `ReverseRegistrar` contract to emit `ReverseClaimed` for testing. 
4. I have added the `testFuzz_SetENSName_Event` foundry test for ERC721PoolFactory. 
5. I have added  the `testFuzz_ReverseClaimed_Event` foundry test for ERC721Pool. 

Please review the changes and let me know if you have any feedback.